File: LLM-Translations/GPT-4/codenet/C/Go/s977214798.go

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func nextInt(sc *bufio.Scanner) int {
	sc.Scan()
	num, _ := strconv.Atoi(sc.Text())
	return num
}

func p(a, b, n int) int {
	if a < b {
		return a + b*n
	} else {
		return b + a*n
	}
}

func f(from int, d []int, e1 []int, e2 []int, b []bool) int {
	if b[from] {
		return -1
	}
	if d[from] != 0 {
		return d[from]
	}
	b[from] = true
	defer func() { b[from] = false }()
	var max int
	to := e1[from]
	if to != 0 {
		now := f(to, d, e1, e2, b)
		if now < 0 {
			return now
		}
		if now > max {
			max = now
		}
	}
	to = e2[from]
	if to != 0 {
		now := f(to, d, e1, e2, b)
		if now < 0 {
			return now
		}
		if now > max {
			max = now
		}
	}
	d[from] = max + 1
	return d[from]
}

func main() {
	sc := bufio.NewScanner(os.Stdin)
	sc.Split(bufio.ScanWords)
	n := nextInt(sc)
	a := make([][]int, n)
	for i := 0; i < n; i++ {
		a[i] = make([]int, n-1)
		for j := 0; j < n-1; j++ {
			a[i][j] = nextInt(sc) - 1
		}
	}
	e1 := make([]int, 1000000)
	e2 := make([]int, 1000000)
	d := make([]int, 1000000)
	b := make([]bool, 1000000)
	for i := 0; i < n; i++ {
		for j := 0; j < n-2; j++ {
			from := p(i, a[i][j], n)
			to := p(i, a[i][j+1], n)
			if e1[from] == 0 {
				e1[from] = to
			} else if e2[from] == 0 {
				e2[from] = to
			} else {
				fmt.Println("-1")
				return
			}
		}
	}
	ans := 0
	for i := 0; i < n; i++ {
		from := p(i, a[i][0], n)
		now := f(from, d, e1, e2, b)
		if now < 0 {
			ans = -1
			break
		}
		if now > ans {
			ans = now
		}
	}
	fmt.Println(ans)
}

File: LLM-Translations/GPT-4/codenet/C/Python/s977214798.py
def nextint(s):
    num = ''
    sign = 1
    while len(s) > 0:
        c = s.pop(0)
        if c == '-':
            sign = -1
        elif c.isdigit():
            num += c
        elif num != '':
            break
    return sign * int(num)

def p(a, b, n):
    if a < b:
        return a + b * n
    else:
        return b + a * n

def f(from_, e1, e2, d, b):
    if b[from_]:
        return -1
    if d[from_]:
        return d[from_]
    b[from_] = True
    max_ = 0
    for to in (e1[from_], e2[from_]):
        if to != 0:
            now = f(to, e1, e2, d, b)
            if now < 0:
                return now
            if now > max_:
                max_ = now
    b[from_] = False
    d[from_] = max_ + 1
    return d[from_]

def main():
    s = list(input().strip())
    n = nextint(s)
    a = [[nextint(s) - 1 for _ in range(n - 1)] for _ in range(n)]
    e1 = [0] * (n * n)
    e2 = [0] * (n * n)
    d = [0] * (n * n)
    b = [False] * (n * n)

    for i in range(n):
        for j in range(n - 2):
            from_ = p(i, a[i][j], n)
            to = p(i, a[i][j + 1], n)
            if e1[from_] == 0:
                e1[from_] = to
            elif e2[from_] == 0:
                e2[from_] = to
            else:
                print(-1)
                return

    ans = 0
    for i in range(n):
        from_ = p(i, a[i][0], n)
        now = f(from_, e1, e2, d, b)
        if now < 0:
            ans = -1
            break
        if now > ans:
            ans = now

    print(ans)

main()


